solr/core/src/java/org/apache/solr/update/processor/NestedUpdateProcessorFactory.java

```diff
@@ -0,0 +1,126 @@
+
+import java.io.IOException;
+import java.util.Collection;
+
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.SolrInputField;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.schema.IndexSchema;
+import org.apache.solr.update.AddUpdateCommand;
+
+public class NestedUpdateProcessorFactory extends UpdateRequestProcessorFactory {
+
+  public UpdateRequestProcessor getInstance(SolrQueryRequest req, SolrQueryResponse rsp, UpdateRequestProcessor next ) {
+    boolean storeParent = shouldStoreDocParent(req.getSchema());
+    boolean storePath = shouldStoreDocPath(req.getSchema());
+    if(!(storeParent || storePath)) {
+      return next;
+    }
+    return new NestedUpdateProcessor(req, rsp, shouldStoreDocParent(req.getSchema()), shouldStoreDocPath(req.getSchema()), next);
+  }
+
+  private static boolean shouldStoreDocParent(IndexSchema schema) {
+    return schema.getFields().containsKey(IndexSchema.NEST_PARENT_FIELD_NAME);
+  }
+
+  private static boolean shouldStoreDocPath(IndexSchema schema) {
+    return schema.getFields().containsKey(IndexSchema.NEST_PATH_FIELD_NAME);
+  }
+}
+
+class NestedUpdateProcessor extends UpdateRequestProcessor {
+  private static final String PATH_SEP_CHAR = "/";
+  private static final String NUM_SEP_CHAR = ",";
+  private static final String SINGLE_VALUE_CHAR = "s";
+  private boolean storePath;
+  private boolean storeParent;
+  private String uniqueKeyFieldName;
+
+
+  protected NestedUpdateProcessor(SolrQueryRequest req, SolrQueryResponse rsp, boolean storeParent, boolean storePath, UpdateRequestProcessor next) {
+    super(next);
+    this.storeParent = storeParent;
+    this.storePath = storePath;
+    this.uniqueKeyFieldName = req.getSchema().getUniqueKeyField().getName();
+  }
+
+  @Override
+  public void processAdd(AddUpdateCommand cmd) throws IOException {
+    SolrInputDocument doc = cmd.getSolrInputDocument();
+    processDocChildren(doc, null);
+    super.processAdd(cmd);
+  }
+
+  private void processDocChildren(SolrInputDocument doc, String fullPath) {
+    int childNum = 0;
+    for(SolrInputField field: doc.values()) {
+      boolean isSingleVal = !(field.getValue() instanceof Collection);
+      for(Object val: field) {
+        if(!(val instanceof SolrInputDocument)) {
+          // either all collection items are child docs or none are.
+          break;
+        }
+        final String fieldName = field.getName();
+
+        if(fieldName.contains(PATH_SEP_CHAR)) {
+          throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Field name: '" + fieldName
+              + "' contains: '" + PATH_SEP_CHAR + "' , which is reserved for the nested URP");
+        }
+        final String sChildNum = isSingleVal ? SINGLE_VALUE_CHAR: String.valueOf(childNum);
+        final String lastPath = fieldName + NUM_SEP_CHAR + sChildNum + NUM_SEP_CHAR;
```

Comment:
I'm a bit confused why SINGLE_VALUE_CHAR was introduced.  Is this so that a DocTransformer can differentiates between a one-ary array and a single doc?  I suppose so.  The choice of 's' was confusing as I thought it was an attempt to pluralize something at first but now I think you intended an abbreviation for "singular".  Could we simply use the empty string in this case?

Also, please define lastPath & jointPath later closer to when they are used, which is immediately prior to calling processChildDoc.

Why the trailing NUM_SEP_CHAR in lastPath?

Reply:
The trailing num sep char is there so we can query all documents in the specified _NEST_PATH_: path#*#